hs_access_control/models.py

```diff
@@ -669,22 +699,91 @@ def can_share_group(self, this_group, this_privilege):
         if __debug__:  # during testing only, check argument types and preconditions
             assert isinstance(this_group, Group)
             assert this_privilege >= PrivilegeCodes.OWNER and this_privilege <= PrivilegeCodes.VIEW
+            if user is not None:
+                assert isinstance(user, User)
 
         if not self.user.is_active: raise PermissionDenied("Requesting user is not active")
         if not this_group.gaccess.active: raise PermissionDenied("Group is not active")
+        if user is not None:
+            if not user.is_active: raise PermissionDenied("Grantee user is not active")
 
         access_group = this_group.gaccess
 
+        # access control logic: 
+        # Can augment privilege if
+        #   Admin
+        #   Owner
+        #   Permission for self 
+        #   Non-owner and shareable
+        # Can downgrade privilege if 
+        #   Admin
+        #   Owner  
+        #   Permission for self
+        # cannot downgrade privilege just by having sharing privilege. 
+      
+        grantor_priv = access_group.get_effective_privilege(self.user)
+        if user is not None:
+            grantee_priv = access_group.get_effective_privilege(user)
+
+        # check for user authorization
         if self.user.is_superuser:
-            return True
+            pass  # admin can do anything
 
-        if not self.owns_group(this_group) and not access_group.shareable:
-            return False   # share raises PermissionDenied("User is not group owner and group is not shareable")
+        elif grantor_priv == PrivilegeCodes.OWNER:
+            pass  # owner can do anything
 
-        # must have a this_privilege greater than or equal to what we want to assign
-        return UserGroupPrivilege.objects.filter(group=this_group,
-                                                 user=self.user,
-                                                 privilege__lte=this_privilege).exists()
+        elif access_group.shareable:
+            if grantor_priv > PrivilegeCodes.VIEW:
+                return False  # raise PermissionDenied("User has no privilege over group")
```

Comment:
Same comment as above. Can do that in check_share_*_with_* and then catch it in can_share_*_with_* and not catch it in share_*_with_*.  Would rather not mess with ui calls here. The intent of can_share_*_with_* is a simple check of whether something can be done.  If people are using it that way, the code will break.